Create guide for adding samples (#659)
microsoft/Quantum: samples/azure-quantum/CONTRIBUTING.md

> # Process for creating notebook samples for the Azure Quantum portal experience

Reviewer: ```suggestion
# Process for contributing notebook samples to the Azure Quantum portal experience

```

> # Process for creating notebook samples for the Azure Quantum portal experience

Reviewer: In general, this file has a lot of Markdownlint issues that should be resolved — I've made some concrete comments with suggestions for some of them, but it'd be worth making sure we don't have any warnings before merging in. Thanks!

> This document focuses on the process for creating samples that will eventually appear in the hosted notebooks sample gallery in the portal. If you are creating a sample for other purposes or means of consumption, then this document might not apply to your scenario. (If you are interested in contributing an optimization notebook sample, please see [this repository](https://github.com/microsoft/qio-samples)).

Reviewer: ```suggestion
This document focuses on the process for creating samples that will eventually appear in the hosted notebooks sample gallery in the portal. If you are creating a sample for other purposes or means of consumption, then this document might not apply to your scenario. (If you are interested in contributing an optimization notebook sample, please see [the microsoft/qio-samples repository](https://github.com/microsoft/qio-samples)).
```

> There are three steps involved in creating a new sample.

Reviewer: ```suggestion
There are three steps involved in contributing a new sample.
```

1. Write the notebook sample
1. Test it in the portal

> 1. Merge it into `main` AND into `feature/samples-gallery`

Reviewer: Another important point: making sure all appropriate metadata is there for inclusion in aka.ms/try-qsharp and docs.microsoft.com/samples. For the former, that's mostly binder-index.md and is largely enforced via CI. For the latter, that consists of making sure there's an appropriate YAML header on each README.md file.

It may help to list a prototypical structure for a new notebook sample. E.g.: something roughly like the following.

```
binder-index.md # Should link to subject area and each way of using each sample
samples/
  subject-area/
     README.md # should describe subject area and link to constituent samples
     sample-name/
         README.md # should have YAML header for docs.ms/samples and manifest section
         sample-name.ipynb
         sample-name.csproj # Not yet supported in portal context.
```

> Once it is merged, a service engineer must be notified to add it to the sample gallery. Once done, the sample will be publicly available in the next deployment, usually within a week.

Reviewer: How should external users do this? Or is this step expected to be taken on only by FTEs on behalf of contributors?

> ## Writing the sample

Reviewer: ```suggestion
## Writing the sample

```

There are a few principles to keep in mind when writing samples that are going to be presented in the portal's sample gallery.

> 1. The command to log in to Azure Quantum should be in its own cell.

Reviewer: This may not be the case for all samples — e.g.: those demonstrating simulator usage.

> 1. There should be no empty cells in the notebook.

Reviewer: I take it this refers mostly to the last cell that sometimes gets added to notebooks? If so, we may want to run a set of notebook lints or something as part of CI?

Author: Indeed, and that could be a good mitigation. For context, empty cells break our E2E tests and detract a bit from the polish.

Reviewer: Agreed about the polish, but a bit confused why that breaks e2e; would love to learn more there!

> 1. The notebook must be entirely self-contained; it cannot reference other files on disk.

Reviewer: We should make sure this is clearly understood as being scoped to portal-hosted notebooks. I understand that's implicit as part of the focus of the whole MD file but it'd be good to make that explicit as well.
```suggestion
1. To run in the portal notebook experience, the notebook must be entirely self-contained; it cannot reference other files on disk.
```

## Testing the sample

> Once the sample is written and working locally, you can test it out in the hosted environment in the Azure portal. In order to make it appear in the sample gallery, you must access the portal using a special link:

Reviewer: ```suggestion
Once the sample is written and working locally, you can test it out in the hosted environment in the Azure Portal. In order to make it appear in the sample gallery, you must access the Portal using a special link:
```

> `https://portal.azure.com/?microsoft_azure_quantum_include_notebook_sample_with_url={URL of the raw sample file}`.

Reviewer: We may want a CI bot that offers links of this form, perhaps?

Author: A nice idea, but probably out of scope for this PR.

> Your browser should automatically  URL-encode the URL you paste, so don't worry about special characters.

Reviewer: ```suggestion
Your browser should automatically URL-encode the URL you paste, so don't worry about special characters.
```

> Make sure that the sample is rendering correctly, all cells run successfully, and the login details are auto-populated with the correct information for the workspace.

Reviewer: ```suggestion
Make sure that the sample is rendering correctly, all cells run successfully, and that the login details are auto-populated with the correct information for the workspace.
```

## Creating the PRs

> First, create a PR against `main` with the notebook sample. Once that is merged, you are ready to merge it back into `feature/samples-gallery`. You might not be able to merge the `main` branch entirely into `feature/samples-gallery` due to some divergence they have, so just create a new branch based on `feature/samples-gallery` and check out your new files for the notebook samples from `main`. In this PR, you must include a few details:

Reviewer: Cherry-picking may also be a solution here.

Author: True, but I'll keep the document simple and just mention one option.

1. The title of the sample. This should be around 20 characters at the most.
1. A brief description of the sample following these guidelines:
   - Length target of 85 characters (+/- 15 characters at the most)
   - Should not begin with “This sample” or other boilerplate wording
   - Should not mention the provider (as that is already noted on the tile)
   - Should speak in an implied 2nd-person view (“Run a job” preferred over “You can run a job”)
   - Verbiage should be specific to function and obvious if correlated to other samples in the gallery

>    - Must be approved by a UX designer and PM

Reviewer: Where is this requirement coming from? That's not something we've generally had for samples, and I'm bit concerned with adding more process here.

Author: It's not exactly related to samples, but we have a policy that any copy that appears in the portal must have sign off. Part of the motivation for this is that the copy we used to have in the sample descriptions were written with very different tone/voice and led to a less scannable page. This generally does not include the actual contents of the sample, only the metadata.

> 1. The provider that the sample targets (e.g. IonQ or Quantinuum). If the sample does not target any providers, explicitly indicate this in the PR summary.

Reviewer: Can you please add Quantinuum to cspell.json? Thanks!

### E2E tests

> When the PR is raised, [this pipeline](https://ms-quantum.visualstudio.com/Quantum%20Program/_build?definitionId=589) will be kicked off. It will take a while and can be a little flakey. If need be, it can be manually bypassed if the tests are failing due to reasons unrelated to your change. Note that if you are adding a new sample, it will not be run in this execution, however it will be run before the change is deployed to the portal. Thus, you do not need to make any extra effort to ensure that the notebook sample is covered by the E2E tests.

Reviewer: The results of this pipeline aren't publicly visible, right? We may need to scope this paragraph a bit more as a result.
